Add an ability to load an arbitrary list of Cortex modules

blind-oracle opened the pull request:
Signed-off-by: Igor Novgorodov <igor@novg.net>

**What this PR does**:
Adds an ability to load a comma-separated list of modules instead of a single module or 'all'.
This should be backwards-compatible with the current behaviour.

The moduleManager is extended a bit to keep track of already loaded modules.
I don't really like having **All** as a module, but I'll leave it for now as-is to keep changes compact.
And single-binary detection depends on this in a couple of places.
Ideally we should just alias the word **all** to a predefined list of modules and load them like they were specified on the command line or in config file.

**Which issue(s) this PR fixes**:
Fixes #3272

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

pracucci commented on CHANGELOG.md:
```diff
@@ -2,6 +2,9 @@
 
 ## master / unreleased
 
+* [FEATURE] Implement an ability to load multiple Cortex modules using a comma-separated list. #3272
+  - Should be backwards compatible with the current behavior.
+  - Target list like 'all,compactor' can be used to load the single-binary modules plus some other modules.
```
I suggest a little rephrasing to simplify it:

```suggestion
* [ENHANCEMENT] Allow to specify multiple comma-separated Cortex services to `-target` CLI option (or its respective YAML config option). For example, `-target=all,compactor` can be used to start Cortex single-binary with compactor as well. #3272
```

We consider these improvements "enhancement" and not new core features. Please move the changelog entry to the right place (entries are grouped by type).

blind-oracle replied:
Sure, forgot that they're grouped.

pracucci commented on pkg/cortex/modules.go:
```diff
@@ -173,7 +173,7 @@ func (t *Cortex) initDistributorService() (serv services.Service, err error) {
 	// whenever it's not running as an internal dependency (ie. querier or
 	// ruler's dependency)
-	canJoinDistributorsRing := (t.Cfg.Target == All || t.Cfg.Target == Distributor)
+	canJoinDistributorsRing := t.Cfg.Modules[Distributor] || t.Cfg.Modules[All]
```
You should do a similar change in `initAlertManager()` too.

pracucci commented on pkg/cortex/cortex.go:
```diff
@@ -71,10 +72,11 @@ import (
-	PrintConfig bool   `yaml:"-"`
-	HTTPPrefix  string `yaml:"http_prefix"`
+	Target      string          `yaml:"target"`
+	Modules     map[string]bool `yaml:"-"`
```
I think keeping both `Target string`  and the new `Modules map[string]bool` may lead to bugs in the future. We may forget to check `Modules` instead of `Target`. A better option may be define a CLI flag custom type in `pkg/util/flagext` which is a `[]string` but implements `flag.Value` and `yaml.Marshaller/Unmarshaller` to marshal/unmarshal a comma separated string.

We wouldn't have a map, but checking if a module is within the `[]string` would be as easy as calling `util.StringsContain()`.

What do you think?

blind-oracle replied:
There's already flagext.StringSlice, should I just use it?

pstibrany replied:
`flagext.StringSlice` doesn't support comma-separated argument, only argument with multiple flags. I'm afraid that would be incompatible change in YAML config.

pracucci commented on pkg/util/modules/modules.go:
```diff
@@ -25,6 +25,12 @@ type module struct {
 	modules map[string]*module
+
+	// Modules that are already initialized
+	modulesLoaded map[string]bool
```
I would rename "loaded" into "initialised", which is the naming we use here.

pracucci commented on pkg/util/modules/modules.go:
```diff
@@ -119,6 +132,11 @@ func (m *Manager) UserVisibleModuleNames() []string {
 }
 
+// GetServicesMap returns services map
+func (m *Manager) GetServicesMap() map[string]services.Service {
```
We may call it `GetInitialisedServices()` and update the comment accordingly.

pracucci commented on pkg/util/modules/modules.go:
```diff
@@ -67,41 +75,46 @@ func (m *Manager) AddDependency(name string, dependsOn ...string) error {
 
 	for ix, n := range deps {
+		// Skip already loaded modules
+		if m.modulesLoaded[n] {
```
Couldn't we just check `m.servicesMap`?

pstibrany replied:
That wouldn't work... not all modules have a service.

pracucci reviewed (commented):
Thanks for working on this! I left few comments. @pstibrany what do you think?

pstibrany commented on pkg/util/modules/modules.go:
```diff
@@ -67,41 +75,46 @@ func (m *Manager) AddDependency(name string, dependsOn ...string) error {
-func (m *Manager) InitModuleServices(target string) (map[string]services.Service, error) {
-	if _, ok := m.modules[target]; !ok {
-		return nil, fmt.Errorf("unrecognised module name: %s", target)
+func (m *Manager) InitModuleServices(name string) error {
```
Could we make it work with this signature instead?

```suggestion
func (m *Manager) InitModuleServices(names ...string) (map[string]services.Service, error)
```

There are multiple projects using this system, and this is backwards compatible, while your suggestion isn't. Furthermore it keeps the state of initialization outside of this manager.

blind-oracle replied:
Sure, didn't know that anything else other than Cortex used it.

pstibrany commented on pkg/util/modules/modules.go:
```diff
@@ -131,6 +149,12 @@ func (m *Manager) IsUserVisibleModule(mod string) bool {
 }
 
+// IsModuleRegistered returns true if given module was registered
+func (m *Manager) IsModuleRegistered(mod string) bool {
```
Let's not introduce new public methods to manager, that are not needed externally.

pstibrany reviewed (commented):
Thank you for working on this! I suggest we keep the changes to `modules.Manager` backwards compatible and not introduce new public API to it (it's always easy to add new methods, and difficult to remove them once they get used), since there are several other projects using this other than Cortex.

blind-oracle commented:
Thanks for the comments everyone.
I've tried to take all of them into account and improve accordingly:
* I've created a new type `flagext.StringSliceCSV` which implements serialisation to/from comma-separated string + tests.
I also had to add a workaround for this type to the doc-generator as it considers it a "list of strings".
* Reverted the API of module loader to be backwards-compatible without extra public methods
* Workaround Jaeger name generation as before it was suffixed by Target.
Now it will be suffixed only if there's only one module listed in Target, otherwise it will be just `cortex`. This should be backwards-compatible also.
* Fixed the CHANGELOG

pstibrany commented on pkg/util/modules/modules.go:
```diff
@@ -25,6 +25,12 @@ type module struct {
 	modules map[string]*module
+
+	// Modules that are already initialized
+	modulesInitialized map[string]bool
```
Let’s not introduce more state to the Manager than necessary. In particular `modulesInitialized` and `servicesMap` don’t need to be fields of the manager.

Calling `InitModuleServices` twice should recreate everything from scratch.

blind-oracle replied:
Done

pstibrany reviewed (commented):
Thank you for addressing my feedback! I have left more comments in the spirit of “minimize state”.

blind-oracle commented on pkg/cortex/cortex.go:
```diff
@@ -305,19 +317,27 @@ func (t *Cortex) setupThanosTracing() {
 	}
 
-	serviceMap, err := t.ModuleManager.InitModuleServices(t.Cfg.Target)
+	t.ServiceMap, err = t.ModuleManager.InitModuleServices(t.Cfg.Target...)
```
Makes sense, done

blind-oracle replied:
I guess I didn't get the first time your idea, will try to get rid of it in `cortex` package.

blind-oracle replied:
Yep, for now made initModules() stateless and fixed tests.

pstibrany commented on pkg/util/modules/modules_test.go:
```diff
@@ -39,13 +46,24 @@ func TestDependencies(t *testing.T) {
-	svcs, err = mm.InitModuleServices("service_unknown")
-	assert.Nil(t, svcs)
-	assert.Error(t, err, fmt.Errorf("unrecognised module name: service_unknown"))
+	// Test loading of the module second time (should be noop)
```
It should not be noop -- it should in fact create new services. Can we extend the test to verify that services from first and second invocation of `mm.InitModuleServices("serviceC")` are different?

blind-oracle replied:
Yeah, the noop part is left from the previous stateful ModuleManager. Fixed

pstibrany reviewed (commented):
Thank you a lot for your patience. This is on the right track, few more minor nits.

blind-oracle commented:
Integration test failed, but looking at the logs it looks like some other transient issue.

pstibrany commented on pkg/cortex/cortex_test.go:
```diff
@@ -68,24 +68,28 @@ func TestCortex(t *testing.T) {
-		Target: All,
 	}
 
+	cfg.Target = []string{All, Compactor}
```
We can put this back into the struct.

pstibrany reviewed (approved):
I have left some non-blocking nitpicky comments (mostly to minimize the diff), but LGTM. Thank you!

blind-oracle commented:
Done mentioned fixes

pstibrany reviewed (approved):
Thank you very much!

blind-oracle commented:
There's a conflict with newer PR https://github.com/cortexproject/cortex/pull/3287, I've merged current master and fixed.

blind-oracle commented:
Looks like tests are unstable.
I also occasionally get failed etcd-related tests when running `go test ./...` locally.

But now it's something else:
```
--- FAIL: TestShuffleShardWithCaching (5.03s)
    ring_test.go:790: lifecycler 0 failed: failed to pick tokens in the KV store, ring: test: failed to CAS test
    ring_test.go:834: 6 != 5
FAIL
```

pstibrany commented:
> Looks like tests are unstable.
> I also occasionally get failed etcd-related tests when running `go test ./...` locally.

Yes, this test is flaky. We need to fix it.

pracucci commented on CHANGELOG.md:
```diff
@@ -48,6 +48,7 @@
 * [CHANGE] Renamed `-redis.enable-tls` CLI flag to `-redis.tls-enabled`, and its respective YAML config option from `enable_tls` to `tls_enabled`. #3298
 * [FEATURE] Added support for shuffle-sharding queriers in the query-frontend. When configured (`-frontend.max-queriers-per-tenant` globally, or using per-tenant limit `max_queriers_per_tenant`), each tenants's requests will be handled by different set of queriers. #3113 #3257
 * [FEATURE] Query-frontend: added `compression` config to support results cache with compression. #3217
+* [ENHANCEMENT] Allow to specify multiple comma-separated Cortex services to `-target` CLI option (or its respective YAML config option). For example, `-target=all,compactor` can be used to start Cortex single-binary with compactor as well. #3272
```
```suggestion
* [ENHANCEMENT] Allow to specify multiple comma-separated Cortex services to `-target` CLI option (or its respective YAML config option). For example, `-target=all,compactor` can be used to start Cortex single-binary with compactor as well. #3275
```
```suggestion
* [ENHANCEMENT] Allow to specify multiple comma-separated Cortex services to `-target` CLI option (or its respective YAML config option). For example, `-target=all,compactor` can be used to start Cortex single-binary with compactor as well. #3272
```

pracucci commented on pkg/cortex/cortex_test.go:
```diff
@@ -68,15 +69,17 @@ func TestCortex(t *testing.T) {
+	serviceMap, err := c.ModuleManager.InitModuleServices(cfg.Target...)
 	require.NoError(t, err)
 	require.NotNil(t, serviceMap)
+	log.Println(serviceMap)
```
I would remove this debug line.

```suggestion
```

pracucci reviewed (approved):
You did great! LGTM! Will merge as soon as the CI passes.
